AEP/aep-00004.md

```diff
@@ -0,0 +1,163 @@
+
+MongoDB proposes two types of shard keys :
+- Hashed keys, which are better at evenly distribute data among shards but worse for multiple-documents read operations, as they likely distribute documents with close shard key value on different shards. (see [Hashed Sharding documentation](https://www.mongodb.com/docs/manual/core/hashed-sharding/))
+- Range keys, which are not ideal for evenly distribute data among shards but better for multiple-documents read operations. (see [Ranged Sharding documentation](https://www.mongodb.com/docs/manual/core/ranged-sharding/))
+
+### Shard the collections
+That being said, you can now shard the collections you want.
+Let's say that inside the "our_database" database, you want to shard the collection named "collection_1", with a hashed shard key on the "_id" field:
+```javascript
+sh.shardCollection("our_database.collection_1", {_id: "hashed})
+```
+
+For more information on sharding possibility, check out the [sh.shardCollection methode documentation](https://www.mongodb.com/docs/manual/reference/method/sh.shardCollection/#mongodb-method-sh.shardCollection)
+
+### Reference
+For more information on how sharding works in MongoDB, check out their [official sharding documentation](https://www.mongodb.com/docs/manual/sharding/)
+
+# Backwards Compatibility
+
+## With ArmoniK Infrastructure (IMPORTANT)
+Currently, the main concern on backwards compatibility is about data persistency in MongoDB, as we haven't focused on the possibility to pass down custom persistent volumes to the Helm release resource. 
+
+Which means that if your database contains data, if you want to upgrade the infrastructure version to deploy MongoDB (sharded or not doesn't matter) with a Helm chart, one of these two situations will :
+- If you are using Persistent Volumes with a reclaim policy set to "Retain", you will be able to keep your data but unable to automatically provide it to the new MongoDB pods freshly created by the Helm release.
+- Your data will be lost if you are using Persistent Volumes with a reclaim policy set to "Delete", which is the default value with dynamic provisioning.
+
+
+## With ArmoniK
+This refactoring would not be backward-compatible with previous versions of ArmoniK since we would prefer to not set default values for the Helm chart version. Which means that calls to this refactored MongoDB module must define the `helm_chart_version` parameter, which is not implemented in previous versions of ArmoniK.
+
+## With ArmoniK Core
+This refactoring would be backward-compatible with previous versions of ArmoniK Core but would not be optimal in case of a sharded MongoDB instance, since these previous versions do not implement the database commands to enable sharding. 
+
+# Security Implications
+The unsharded Bitnami's MongoDB Helm chart does not generate TLS certificate for clients. The sharded one does not seem to support TLS at all.
```

Comment:
This is a very big problem but here we should probably mention that our deployement are only example and could be/should be different in a production environment  and respect the security policies of each company.

Reply:
You are talking about the sharded Helm chart not supporting TLS ?